src/rust/bitbox02-rust-c/src/util.rs

```diff
@@ -119,46 +115,36 @@ impl AsRef<str> for CStrMut {
     /// with 0 length due to limitation in `core::slice`.
     fn as_ref(&self) -> &str {
         let mut buf = self.buf;
-        if self.len == 0 && self.buf.is_null() {
+        if self.cap == 0 && self.buf.is_null() {
             buf = core::ptr::NonNull::dangling().as_ptr();
         }
         assert!(!buf.is_null());
 
-        // Find null terminator
-        let mut len = 0;
-        unsafe {
-            let mut b = buf;
-            while b.read() != 0 {
-                b = b.offset(1);
-                len += 1;
-                if len == self.len {
-                    panic!("CStrMut not null terminated");
-                }
-            }
-        }
-
-        unsafe { core::str::from_utf8_unchecked(core::slice::from_raw_parts(buf, len)) }
+        unsafe { core::str::from_utf8_unchecked(core::slice::from_raw_parts(self.buf, self.len)) }
     }
 }
 
 impl CStrMut {
-    /// Create a slice to a buffer. Only allowed for non-null pointers with length or null pointers
-    /// with 0 length due to limitation in `core::slice`.
-    /// The caller must ensure that the string is null terminated.
-    pub unsafe fn as_bytes_mut(&mut self) -> &mut [u8] {
-        let mut buf = self.buf;
-        if self.len == 0 && self.buf.is_null() {
-            buf = core::ptr::NonNull::dangling().as_ptr();
+    /// Create a mutable slice to an unused range of the buffer. Will ensure that byte immediately
+    /// following the range is null.
+    pub fn unused(&mut self, req: usize) -> &mut [u8] {
+        // Must be room for requested amount of bytes and null terminator.
+        if self.cap - self.len < req + 1 {
+            panic!("Not enough bytes left in buffer");
+        }
+        let len = self.len;
+        self.len += req;
```

Comment:
i'm not sure I see the purpose of being able to call `unused()` multiple times to do a sort of `concat`. Do you have a specific use case for it?

Reply:
I was trying to come up with a safe interface so that the business logic code doesn't have to do null termination and use unsafe functions

Reply:
not sure how that line is related to it, this is ensured by the checks below. 
is calling `.write()` twice expected to not overwrite, but append? if so it would be good to document.

Reply:
yes, exactly, if a container is growable it will append by defualt:

```rust
use std::fmt::Write;
fn main() {
    let mut s = String::from("abc");
    let _ = write!(s, "d");
    assert!(&s == "abcd");
}
```

Reply:
oh, and regarding your first comment. If you ever write over bytes in a string you also have to ensure that the following 3 bytes are valid. (So that you are not eating into a 4 byte utf-8 char by mistake). So anything that is appending is much easier than something that is modifying.